uc-crux-llvm/src/UCCrux/LLVM/Run/Simulate.hs

```diff
@@ -0,0 +1,221 @@
+
+import qualified What4.Interface as What4
+import qualified What4.ProgramLoc as What4
+import qualified What4.FunctionName as What4
+
+-- crucible
+import qualified Lang.Crucible.CFG.Core as Crucible
+import qualified Lang.Crucible.FunctionHandle as Crucible
+import qualified Lang.Crucible.Backend as Crucible
+import qualified Lang.Crucible.Simulator as Crucible
+import qualified Lang.Crucible.Types as CrucibleTypes
+
+-- crucible-llvm
+import           Lang.Crucible.LLVM( llvmGlobals )
+import           Lang.Crucible.LLVM.MemModel (MemOptions,  LLVMAnnMap)
+import           Lang.Crucible.LLVM.Translation (transContext, llvmTypeCtx)
+
+import           Lang.Crucible.LLVM.MemModel.Partial (BoolAnn(BoolAnn))
+import           Lang.Crucible.LLVM.Extension( LLVM )
+
+-- crux
+import qualified Crux
+
+import           Crux.Config.Common (CruxOptions)
+import           Crux.Log (outputHandle, OutputConfig(..))
+
+ -- crux-llvm
+import           Crux.LLVM.Overrides (ArchOk)
+import           Crux.LLVM.Simulate (setupSimCtxt, registerFunctions)
+
+ -- local
+import           UCCrux.LLVM.Classify (classifyAssertion, classifyBadBehavior, Explanation(..), Uncertainty(..))
+import           UCCrux.LLVM.Constraints (Constraints, argConstraints, globalConstraints, relationalConstraints)
+import           UCCrux.LLVM.Context.App (AppContext, log)
+import           UCCrux.LLVM.Context.Function (FunctionContext, functionName)
+import           UCCrux.LLVM.Context.Module (ModuleContext, llvmModule, moduleTranslation)
+import           UCCrux.LLVM.Errors.Panic (panic)
+import           UCCrux.LLVM.Logging (Verbosity(Hi))
+import           UCCrux.LLVM.FullType (MapToCrucibleType)
+import           UCCrux.LLVM.PP (ppRegMap)
+import           UCCrux.LLVM.Setup (setupExecution, SetupResult(SetupResult), SetupAssumption(SetupAssumption))
+import           UCCrux.LLVM.Setup.Monad (ppSetupError)
+{- ORMOLU_ENABLE -}
+
+simulateLLVM ::
+  ArchOk arch =>
+  AppContext ->
+  ModuleContext arch ->
+  FunctionContext m arch argTypes ->
+  Crucible.HandleAllocator ->
+  IORef [Explanation m arch argTypes] ->
+  Constraints m argTypes ->
+  Crucible.CFG LLVM blocks (MapToCrucibleType arch argTypes) ret ->
+  MemOptions ->
+  Crux.SimulatorCallback
+simulateLLVM appCtx modCtx funCtx halloc explRef constraints cfg memOptions =
+  Crux.SimulatorCallback $ \sym _maybeOnline ->
+    do
+      let trans = modCtx ^. moduleTranslation
+      let llvmCtxt = trans ^. transContext
+      bbMapRef <- newIORef (Map.empty :: LLVMAnnMap sym)
+      let ?lc = llvmCtxt ^. llvmTypeCtx
+      let ?recordLLVMAnnotation = \an bb -> modifyIORef bbMapRef (Map.insert an bb)
+      let simctx =
+            (setupSimCtxt halloc sym memOptions llvmCtxt)
+              { Crucible.printHandle = view outputHandle ?outputConfig
+              }
+
+      unless (0 == Map.size (constraints ^. globalConstraints)) $
+        panic "simulateLLVM" ["Unimplemented: global constraints"]
```

Comment:
This is really nitpicking, but `Map` also has a `null` predicate

Reply:
TIL! Fixed.